File: benchmarks/matmul/c/matmul.c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char **argv) {
    int n = argc > 1 ? atoi(argv[1]) : 400;
    double *a = malloc((size_t)n * n * sizeof(double));
    double *b = malloc((size_t)n * n * sizeof(double));
    double *c = calloc((size_t)n * n, sizeof(double));

    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) {
            a[i * n + j] = (double)((i * 3 + j) % 13);
            b[i * n + j] = (double)((i + j * 2) % 17);
        }

    for (int i = 0; i < n; i++)
        for (int k = 0; k < n; k++) {
            double aik = a[i * n + k];
            for (int j = 0; j < n; j++)
                c[i * n + j] += aik * b[k * n + j];
        }

    double expected = 0.0;
    for (int k = 0; k < n; k++)
        expected += a[k] * b[k * n];
    if (c[0] != expected) {
        fprintf(stderr, "self-check failed: c[0][0] mismatch\n");
        return 1;
    }

    printf("%.2f\n", c[(size_t)(n - 1) * n + (n - 1)]);
    free(a);
    free(b);
    free(c);
    return 0;
}

File: benchmarks/matmul/python/matmul.py
import sys


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 400

    a = [[(i * 3 + j) % 13 for j in range(n)] for i in range(n)]
    b = [[(i + j * 2) % 17 for j in range(n)] for i in range(n)]
    c = [[0] * n for _ in range(n)]

    for i in range(n):
        ai = a[i]
        ci = c[i]
        for k in range(n):
            aik = ai[k]
            bk = b[k]
            for j in range(n):
                ci[j] += aik * bk[j]

    expected = sum(a[0][k] * b[k][0] for k in range(n))
    assert c[0][0] == expected, "self-check failed: c[0][0] mismatch"
    print(f"{c[n - 1][n - 1]:.2f}")


if __name__ == "__main__":
    main()
